src/main/diagnostics/descriptorActionProvider.ts

```diff
@@ -21,52 +23,99 @@ export class DescriptorActionProvider extends AbstractFileActionProvider {
                     issue.impactedTree?.children
                         ?.map(impact => impact.name)
-                        .forEach(directDependencyId => {
-                            if (proceesedDependencies.has(directDependencyId)) {
-                                // Direct dependency already had a diffrent dependency with issue that generated information for it
-                                return;
-                            }
-                            // Create diagnostics for the dependency
-                            proceesedDependencies.add(directDependencyId);
-                            this.generateInformation(
+                        .forEach(directDependencyId =>
+                            this.handleIssueInDirectDependencyDiagnostic(
+                                issue,
                                 directDependencyId,
                                 dependencyWithIssue.type,
-                                dependencyWithIssue.severity,
-                                document,
-                                textEditor,
-                                diagnostics
-                            );
-                        });
+                                proceesedDependencies,
+                                diagnostics,
+                                document
+                            )
+                        );
                 });
             });
+            // Add gutter icons for top severity and set diagnostics in collection
+            Array.from(proceesedDependencies.values()).forEach(info => {
+                this.addGutter(textEditor, SeverityUtils.getIcon(info.severity), info.positions);
+            });
             this._diagnosticCollection.set(document.uri, diagnostics);
         }
     }
 
-    private generateInformation(
+    /**
+     * Create diagnostics for issue if not exists for a dependency
+     * @param issue - the issue to create diagnosics for
+     * @param directDependencyId  - the direct dependency id to create diagnosics for
+     * @param packgeType - the direct dependency packge type
+     * @param proceesedDependencies - the list of all the processed dependecies to search inside
+     * @param diagnostics - list of all the diagnostics of the document
+     * @param document - the document that holds the dependency
+     */
+    private handleIssueInDirectDependencyDiagnostic(
+        issue: IssueTreeNode,
         directDependencyId: string,
         packgeType: PackageType,
-        severity: Severity,
-        document: vscode.TextDocument,
-        textEditor: vscode.TextEditor,
-        diagnostics: vscode.Diagnostic[]
+        proceesedDependencies: Map<string, InfoPair>,
+        diagnostics: vscode.Diagnostic[],
+        document: vscode.TextDocument
     ) {
+        // Get/create the information
+        let info: InfoPair | undefined = this.getOrCreateDirectDependencyInfo(directDependencyId, packgeType, proceesedDependencies, document);
```

Comment:
There seems to be a problem where the function named get or create may return

Reply:
renamed